jax/experimental/jax2tf/call_tf.py

```diff
@@ -0,0 +1,232 @@
+                lambda x, y: x + y, accumulator, scaled_dres_darg))
+
+      tf.nest.map_structure(acc_ct, res, ct_res)
+      return accumulator
+    # Use call_tf to call the VJP function
+    return call_tf(tf_vjp_fun)(args_jax, ct_res)
+
+  make_call.defvjp(make_call_vjp_fwd, make_call_vjp_bwd)
+  return util.wraps(func_tf)(make_call)
+
+
+call_tf_p = core.Primitive("call_tf")
+call_tf_p.multiple_results = True
+
+
+# The impl will be used in op-by-op mode and calls func_tf in TF eager mode.
+def _call_tf_impl(*args, args_treedef, func_tf, **_):
+  # TODO(necula): does this copy to the host?
+  args_tf_flat = [tf.constant(np.asarray(a)) for a in args]
+  res_tf = func_tf(*args_treedef.unflatten(args_tf_flat))
+  # TODO(necula): check the result for tree and aval
+  # TODO(necula): does this copy to the host?
+  return [jnp.asarray(np.asarray(r)) for r in tree_util.tree_flatten(res_tf)[0]]
+
+
+call_tf_p.def_impl(_call_tf_impl)
+
+
+def _call_tf_abstract_eval(*_, res_tf_sig_flat, **__):
+  return tuple([
+      core.ShapedArray(np.shape(r), _to_jax_dtype(r.dtype))
+      for r in res_tf_sig_flat
+  ])
+
+
+call_tf_p.def_abstract_eval(_call_tf_abstract_eval)
+
+
+def _call_tf_translation_rule(builder, *args_op, func_tf,
+                              args_treedef, args_tf_sig_flat, res_tf_sig_flat,
+                              **_):
+  # TODO(necula): It seems that we need concrete tensors for get_compiler_ir?
+  args_tf_flat = [
+      tf.constant((0 if a.dtype != tf.bool else False),
+                  shape=a.shape,
+                  dtype=a.dtype) for a in args_tf_sig_flat
+  ]
+  args_tf = args_treedef.unflatten(args_tf_flat)
+  func_tf = tf.function(func_tf, jit_compile=True)
+  func_tf_concrete = func_tf.get_concrete_function(*args_tf)
+  captured_ops = []  # Same order as captured_inputs
+  if func_tf_concrete.captured_inputs:
+    # The function uses either captured variables or tensors.
+    msg = (
+      "call_tf works best with a TensorFlow function that does not capture "
+      "variables or tensors from the context. "
+      "See https://github.com/google/jax/blob/master/jax/experimental/jax2tf/README.md#calling-tensorflow-functions-from-jax for a discussion. "
+      f"The following captures were found {func_tf_concrete.captured_inputs}")
+    logging.warning(msg)
+
+    next_var_idx = 0
+    for inp in func_tf_concrete.captured_inputs:
+      if inp.dtype == tf.resource:  # A variable; assume the next variable
+        assert next_var_idx < len(func_tf_concrete.variables)
+        # TODO(necula): better checking that we are picking the right variable
+        var = func_tf_concrete.variables[next_var_idx]
+        assert inp.shape == var.shape
+        next_var_idx += 1
+        inp_const = np.asarray(var)
+      else:
+        inp_const = np.asarray(inp)
+      captured_ops.append(xops.ConstantLiteral(builder, np.asarray(inp_const)))
+
+  # For unoptimized HLO, does not make a different which device we use.
```

Comment:
Are we sure TF won't do similar things to JAX and stage out platform specific implementations of certain things?

Reply:
Not yet sure. I will turn it into a TODO. I plan to run the 2300+ primitive harness through jax2tf and then back with call_tf to see if that uncovers some surprises.